dan_layer/template_lib/src/models/non_fungible.rs

```diff
@@ -263,20 +281,41 @@ impl FromStr for NonFungibleAddress {
     type Err = ParseNonFungibleAddressError;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        // the expected format is "resource_xxxx nft_xxxxx"
-        match s.split_once(' ') {
-            Some((resource_str, addr_str)) => match addr_str.split_once('_') {
-                Some(("nft", id_str)) => {
-                    let resource_addr = ResourceAddress::from_str(resource_str)
-                        .map_err(|e| ParseNonFungibleAddressError::InvalidResource(e.to_string()))?;
-                    let id = NonFungibleId::try_from_canonical_string(id_str)
-                        .map_err(ParseNonFungibleAddressError::InvalidId)?;
-                    Ok(NonFungibleAddress::new(resource_addr, id))
-                },
-                _ => Err(ParseNonFungibleAddressError::InvalidFormat),
-            },
-            None => Err(ParseNonFungibleAddressError::InvalidFormat),
+        // the expected format is nft_{resource_hex}_{type}_{id}
+        let mut splitted = s.split('_');
+
+        if let Some(token) = splitted.next() {
+            if token != "nft" {
+                return Err(ParseNonFungibleAddressError::InvalidFormat);
+            }
+        } else {
+            return Err(ParseNonFungibleAddressError::InvalidFormat);
         }
+
+        let resource_addr: ResourceAddress;
+        if let Some(token) = splitted.next() {
+            resource_addr = ResourceAddress::from_str(token)
+                .map_err(|e| ParseNonFungibleAddressError::InvalidResource(e.to_string()))?;
+        } else {
+            return Err(ParseNonFungibleAddressError::InvalidFormat);
+        }
+
+        let id_type: &str;
+        if let Some(token) = splitted.next() {
+            id_type = token;
+        } else {
+            return Err(ParseNonFungibleAddressError::InvalidFormat);
+        }
+
+        let id: NonFungibleId;
+        if let Some(token) = splitted.next() {
+            id = NonFungibleId::try_from_canonical_string(&format!("{id_type}_{token}"))
+                .map_err(ParseNonFungibleAddressError::InvalidId)?;
+        } else {
+            return Err(ParseNonFungibleAddressError::InvalidFormat);
+        }
+
+        Ok(NonFungibleAddress::new(resource_addr, id))
```

Comment:
Why different parsing code here?

```suggestion
        // the expected format is nft_{resource_hex}_{type}_{id}
let resp = s.strip_prefix("nft_").unwrap_or(s);
 
                let (resource, nft_rest) = rest
                    .split_once('_')
                    .ok_or_else(|| ParseNonFungibleError::InvalidFormat))?;

                let resource_addr =
                    ResourceAddress::from_hex(resource).map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;

                let nft_id = NonFungibleId::try_from_canonical_string(nft_rest))
                    .map_err(|_| InvalidSubstateIdFormat(s.to_string()))?;
                                   

        Ok(NonFungibleAddress::new(resource_addr, nft_id))
```

Then in SubstateId from_str impl - use the above

```rust
NonFungibleAddress::from_str(rest).map_err(...)?;
```